File: src/formosa/geomorphology/geometry/_backends/intersections.f90
!> Evaluates geometric predicates using the Fortran backend.
!!
!! This internal module provides orientation and in-circle
!! predicates for triangulation, together with line-segment
!! intersection routines. It is called by the Python geometry API
!! and other Fortran routines and is not intended to be used
!! directly.
!!
!! Last modified: 2026-08-18, En-Chi Lee (williameclee@gmail.com)
module intersections
    use iso_c_binding, only: c_double, c_int8_t, c_int32_t, c_int64_t
    implicit none(type, external)
    private :: incircle_double, saturate_int32, saturate_int64
    interface bboxes_overlap
        module procedure bboxes_overlap_int32
        module procedure bboxes_overlap_real
    end interface bboxes_overlap
    interface orient
        module procedure orient_real
        module procedure orient_int32
        module procedure orient_int64
    end interface orient
    interface xcross_orient
        module procedure xcross_orient_real
        module procedure xcross_orient_int32
        module procedure xcross_orient_int64
    end interface xcross_orient
    interface xcross
        module procedure xcross_real
        module procedure xcross_int32
        module procedure xcross_int64
    end interface xcross
    interface incircle
        module procedure incircle_real
        module procedure incircle_int32
        module procedure incircle_int64
    end interface incircle
contains
    pure function bboxes_overlap_int32(p1, p2, p3, p4) result(flag_overlap)
        implicit none(type, external)
        ! Arguments
        integer(c_int32_t), intent(in) :: p1(2), p2(2), p3(2), p4(2)
        ! Outputs
        logical(kind=1) :: flag_overlap

        flag_overlap = &
            (max(min(p1(1), p2(1)), min(p3(1), p4(1))) <= &
             min(max(p1(1), p2(1)), max(p3(1), p4(1)))) .and. &
            (max(min(p1(2), p2(2)), min(p3(2), p4(2))) <= &
             min(max(p1(2), p2(2)), max(p3(2), p4(2))))
    end function bboxes_overlap_int32

    pure function bboxes_overlap_real(p1, p2, p3, p4) result(flag_overlap)
        implicit none(type, external)
        ! Arguments
        real, intent(in) :: p1(2), p2(2), p3(2), p4(2)
        ! Outputs
        logical(kind=1) :: flag_overlap

        flag_overlap = &
            (max(min(p1(1), p2(1)), min(p3(1), p4(1))) <= &
             min(max(p1(1), p2(1)), max(p3(1), p4(1)))) .and. &
            (max(min(p1(2), p2(2)), min(p3(2), p4(2))) <= &
             min(max(p1(2), p2(2)), max(p3(2), p4(2))))
    end function bboxes_overlap_real

    pure function saturate_int32(value) result(saturated)
        implicit none(type, external)
        real(c_double), intent(in) :: value
        integer(c_int32_t) :: saturated

        if (value > real(huge(saturated), kind=c_double)) then
            saturated = huge(saturated)
        else if (value < -real(huge(saturated), kind=c_double) - 1.0_c_double) then
            saturated = -huge(saturated) - 1_c_int32_t
        else
            saturated = int(value, kind=c_int32_t)
        end if
    end function saturate_int32

    pure function saturate_int64(value) result(saturated)
        implicit none(type, external)
        real(c_double), intent(in) :: value
        integer(c_int64_t) :: saturated

        ! Converting INT64_MAX to double rounds upward to 2**63, so
        ! use inclusive comparisons before conversion to keep it
        ! defined.
        if (value >= real(huge(saturated), kind=c_double)) then
            saturated = huge(saturated)
        else if (value <= -real(huge(saturated), kind=c_double)) then
            saturated = -huge(saturated) - 1_c_int64_t
        else
            saturated = int(value, kind=c_int64_t)
        end if
    end function saturate_int64

    !> Calculates the signed orientation determinant for real
    !! coordinates.
    !!
    !! A positive result denotes counterclockwise orientation. A
    !! negative result denotes clockwise orientation. Zero denotes
    !! collinearity.
    pure function orient_real(p1, p2, p3) result(o)
        implicit none(type, external)
        ! Arguments
        real, intent(in) :: p1(2), p2(2), p3(2)
        ! Outputs
        real :: o

        o = (p2(1) - p1(1))*(p3(2) - p1(2)) - (p2(2) - p1(2))*(p3(1) - p1(1))
    end function orient_real

    !> Calculates the signed orientation determinant for 32-bit
    !! coordinates.
    !!
    !! The calculation uses double-precision intermediates. A result
    !! outside the 32-bit range is saturated to the corresponding
    !! integer limit. The result retains the 32-bit input kind and
    !! preserves the predicate sign.
    pure function orient_int32(p1, p2, p3) result(o)
        implicit none(type, external)
        ! Arguments
        integer(c_int32_t), intent(in) :: p1(2), p2(2), p3(2)
        ! Outputs
        integer(c_int32_t) :: o
        ! Local variables
        real(c_double) :: p1d(2), p2d(2), p3d(2)

        p1d = real(p1, kind=c_double)
        p2d = real(p2, kind=c_double)
        p3d = real(p3, kind=c_double)
        o = saturate_int32((p2d(1) - p1d(1))*(p3d(2) - p1d(2)) - &
                           (p2d(2) - p1d(2))*(p3d(1) - p1d(1)))
    end function orient_int32

    !> Calculates the signed orientation determinant for 64-bit
    !! coordinates.
    !!
    !! The calculation uses double-precision intermediates. A result
    !! outside the 64-bit range is saturated to the corresponding
    !! integer limit. The result retains the 64-bit input kind and
    !! preserves the predicate sign.
    pure function orient_int64(p1, p2, p3) result(o)
        implicit none(type, external)
        ! Arguments
        ! f2py does not currently emit its long_long typedef for
        ! c_int64_t.
        !f2py integer(kind=8), intent(in) :: p1, p2, p3
        integer(c_int64_t), intent(in) :: p1(2), p2(2), p3(2)
        ! Outputs
        !f2py integer(kind=8) :: o
        integer(c_int64_t) :: o
        ! Local variables
        real(c_double) :: p1d(2), p2d(2), p3d(2)

        p1d = real(p1, kind=c_double)
        p2d = real(p2, kind=c_double)
        p3d = real(p3, kind=c_double)
        o = saturate_int64((p2d(1) - p1d(1))*(p3d(2) - p1d(2)) - &
                           (p2d(2) - p1d(2))*(p3d(1) - p1d(1)))
    end function orient_int64

    !> Calculates the signed in-circle determinant for real
    !! coordinates.
    !!
    !! For counterclockwise triangle vertices, a positive result
    !! places the test point inside their circumcircle. Zero places
    !! the point on the circle, and a negative result places it
    !! outside.
    pure function incircle_real(a, b, c, p) result(det)
        implicit none(type, external)
        ! Arguments
        real, intent(in) :: a(2), b(2), c(2), p(2)
        ! Outputs
        real :: det
        ! Local variables
        real :: adx, ady, bdx, bdy, cdx, cdy
        real :: abdet, bcdet, cadet, alift, blift, clift

        adx = a(1) - p(1)
        ady = a(2) - p(2)
        bdx = b(1) - p(1)
        bdy = b(2) - p(2)
        cdx = c(1) - p(1)
        cdy = c(2) - p(2)

        abdet = adx*bdy - bdx*ady
        bcdet = bdx*cdy - cdx*bdy
        cadet = cdx*ady - adx*cdy
        alift = adx*adx + ady*ady
        blift = bdx*bdx + bdy*bdy
        clift = cdx*cdx + cdy*cdy
        det = alift*bcdet + blift*cadet + clift*abdet
    end function incircle_real

    !> Calculates the signed in-circle determinant for 32-bit
    !! coordinates.
    !!
    !! The calculation uses double-precision intermediates. A result
    !! outside the 32-bit range is saturated to the corresponding
    !! integer limit. The result retains the 32-bit input kind and
    !! preserves the predicate sign.
    pure function incircle_int32(a, b, c, p) result(det)
        implicit none(type, external)
        ! Arguments
        integer(c_int32_t), intent(in) :: a(2), b(2), c(2), p(2)
        ! Outputs
        integer(c_int32_t) :: det

        det = saturate_int32( &
              incircle_double(real(a, kind=c_double), real(b, kind=c_double), &
                              real(c, kind=c_double), real(p, kind=c_double)))
    end function incircle_int32

    !> Calculates the sign of the in-circle determinant for 32-bit
    !! coordinates.
    pure function incircle_pos_int32(a, b, c, p) result(pos)
        implicit none(type, external)
        ! Arguments
        integer(c_int32_t), intent(in) :: a(2), b(2), c(2), p(2)
        ! Outputs
        logical(kind=1) :: pos

        pos = incircle_double( &
              real(a, kind=c_double), real(b, kind=c_double), &
              real(c, kind=c_double), real(p, kind=c_double)) > 0
    end function incircle_pos_int32

    !> Calculates the signed in-circle determinant for 64-bit
    !! coordinates.
    !!
    !! The calculation uses double-precision intermediates. A result
    !! outside the 64-bit range is saturated to the corresponding
    !! integer limit. The result retains the 64-bit input kind and
    !! preserves the predicate sign.
    pure function incircle_int64(a, b, c, p) result(det)
        implicit none(type, external)
        ! Arguments
        !f2py integer(kind=8), intent(in) :: a, b, c, p
        integer(c_int64_t), intent(in) :: a(2), b(2), c(2), p(2)
        ! Outputs
        !f2py integer(kind=8) :: det
        integer(c_int64_t) :: det

        det = saturate_int64( &
              incircle_double(real(a, kind=c_double), real(b, kind=c_double), &
                              real(c, kind=c_double), real(p, kind=c_double)))
    end function incircle_int64

    !> Calculates the signed in-circle determinant.
    pure function incircle_double(a, b, c, p) result(det)
        ! Double-precision implementation shared by the integer
        ! entry points.
        implicit none(type, external)
        ! Arguments
        real(c_double), intent(in) :: a(2), b(2), c(2), p(2)
        ! Outputs
        real(c_double) :: det
        ! Local variables
        real(c_double) :: adx, ady, bdx, bdy, cdx, cdy
        real(c_double) :: abdet, bcdet, cadet, alift, blift, clift

        adx = a(1) - p(1)
        ady = a(2) - p(2)
        bdx = b(1) - p(1)
        bdy = b(2) - p(2)
        cdx = c(1) - p(1)
        cdy = c(2) - p(2)

        abdet = adx*bdy - bdx*ady
        bcdet = bdx*cdy - cdx*bdy
        cadet = cdx*ady - adx*cdy
        alift = adx*adx + ady*ady
        blift = bdx*bdx + bdy*bdy
        clift = cdx*cdx + cdy*cdy
        det = alift*bcdet + blift*cadet + clift*abdet
    end function incircle_double

    pure function on_segment(a, b, p) result(on_flag)
        implicit none(type, external)
        ! Arguments
        real, intent(in) :: a(2), b(2), p(2)
        ! Outputs
        logical :: on_flag

        if (orient_real(a, b, p) /= 0) then
            on_flag = .false.
            return
        end if

        on_flag = p(1) >= min(a(1), b(1)) .and. &
                  p(1) <= max(a(1), b(1)) .and. &
                  p(2) >= min(a(2), b(2)) .and. &
                  p(2) <= max(a(2), b(2))
    end function on_segment

    elemental function xcross_orient_real(o_abc, o_abd, o_cda, o_cdb) &
        result(flag)
        implicit none(type, external)
        ! Arguments
        real, intent(in) :: o_abc, o_abd, o_cda, o_cdb
        logical(kind=1) :: flag

        flag = ((o_abc /= 0) .and. (o_abd /= 0) .and. &
                (o_cda /= 0) .and. (o_cdb /= 0) .and. &
                ((o_abc > 0) .neqv. (o_abd > 0)) .and. &
                ((o_cda > 0) .neqv. (o_cdb > 0)))
    end function xcross_orient_real

    elemental function xcross_orient_int32(o_abc, o_abd, o_cda, o_cdb) &
        result(flag)
        implicit none(type, external)
        ! Arguments
        integer(c_int32_t), intent(in) :: o_abc, o_abd, o_cda, o_cdb
        logical(kind=1) :: flag

        flag = ((o_abc /= 0) .and. (o_abd /= 0) .and. &
                (o_cda /= 0) .and. (o_cdb /= 0) .and. &
                ((o_abc > 0) .neqv. (o_abd > 0)) .and. &
                ((o_cda > 0) .neqv. (o_cdb > 0)))
    end function xcross_orient_int32

    elemental function xcross_orient_int64(o_abc, o_abd, o_cda, o_cdb) &
        result(flag)
        implicit none(type, external)
        ! Arguments
        integer(c_int64_t), intent(in) :: o_abc, o_abd, o_cda, o_cdb
        logical(kind=1) :: flag

        flag = ((o_abc /= 0) .and. (o_abd /= 0) .and. &
                (o_cda /= 0) .and. (o_cdb /= 0) .and. &
                ((o_abc > 0) .neqv. (o_abd > 0)) .and. &
                ((o_cda > 0) .neqv. (o_cdb > 0)))
    end function xcross_orient_int64

    pure function xcross_real(a, b, c, d) result(flag)
        implicit none(type, external)
        ! Arguments
        real, intent(in) :: a(2), b(2), c(2), d(2)
        logical(kind=1) :: flag
        real :: o_abc, o_abd, o_cda, o_cdb

        o_abc = orient(a, b, c)
        o_abd = orient(a, b, d)
        o_cda = orient(c, d, a)
        o_cdb = orient(c, d, b)
        flag = xcross_orient(o_abc, o_abd, o_cda, o_cdb)
    end function xcross_real

    pure function xcross_int32(a, b, c, d) result(flag)
        implicit none(type, external)
        ! Arguments
        integer(c_int32_t), intent(in) :: a(2), b(2), c(2), d(2)
        logical(kind=1) :: flag
        integer(c_int32_t) :: o_abc, o_abd, o_cda, o_cdb

        o_abc = orient(a, b, c)
        o_abd = orient(a, b, d)
        o_cda = orient(c, d, a)
        o_cdb = orient(c, d, b)
        flag = xcross_orient(o_abc, o_abd, o_cda, o_cdb)
    end function xcross_int32

    pure function xcross_int64(a, b, c, d) result(flag)
        implicit none(type, external)
        ! Arguments
        integer(c_int64_t), intent(in) :: a(2), b(2), c(2), d(2)
        logical(kind=1) :: flag
        integer(c_int64_t) :: o_abc, o_abd, o_cda, o_cdb

        o_abc = orient(a, b, c)
        o_abd = orient(a, b, d)
        o_cda = orient(c, d, a)
        o_cdb = orient(c, d, b)
        flag = xcross_orient(o_abc, o_abd, o_cda, o_cdb)
    end function xcross_int64

    !> Classifies the intersection of 2 closed 2D line segments.
    !!
    !! Flags:
    !! - -1 : disjoint
    !! -  0 : endpoint-to-endpoint touch
    !! -  1 : interior-interior crossing (X)
    !! -  2 : collinear overlap, not identical
    !! -  3 : identical segment
    !! -  4 : endpoint-on-interior (T-junction)
    !! -  5 : degenerate segment (some line is actually a point)
    pure function lines_intersect(l1a, l1b, l2a, l2b) result(flag)
        implicit none(type, external)
        ! Arguments
        real, intent(in) :: l1a(2), l1b(2), l2a(2), l2b(2)
        ! Outputs
        integer(c_int8_t) :: flag
        ! Local variables
        logical(kind=1) :: eq_l1al2a, eq_l1al2b, eq_l1bl2a, eq_l1bl2b
        real :: o1, o2, o3, o4
        real :: a0, a1, c0, c1, tmp, overlap0, overlap1

        flag = -1

        if (all(l1a == l1b) .or. all(l2a == l2b)) then ! Degeneracy test
            flag = 5
            return
        else if (.not. bboxes_overlap(l1a, l1b, l2a, l2b)) then ! Bounding box intersection test
            flag = -1
            return
        end if

        eq_l1al2a = all(l1a == l2a)
        eq_l1al2b = all(l1a == l2b)
        eq_l1bl2a = all(l1b == l2a)
        eq_l1bl2b = all(l1b == l2b)
        if ((eq_l1al2a .and. eq_l1bl2b) .or. (eq_l1al2b .and. eq_l1bl2a)) then ! Same line test
            flag = 3
            return
        end if

        ! Interior-interior crossing
        o1 = orient_real(l1a, l1b, l2a)
        o2 = orient_real(l1a, l1b, l2b)
        o3 = orient_real(l2a, l2b, l1a)
        o4 = orient_real(l2a, l2b, l1b)
        if (xcross_orient(o1, o2, o3, o4)) then
            flag = 1
            return
        end if

        ! Collinear case.
        if ((o1 == 0) .and. (o2 == 0) .and. (o3 == 0) .and. (o4 == 0)) then
            ! Project onto the dominant axis.
            if (abs(l1b(1) - l1a(1)) >= abs(l1b(2) - l1a(2))) then
                a0 = l1a(1)
                a1 = l1b(1)
                c0 = l2a(1)
                c1 = l2b(1)
            else
                a0 = l1a(2)
                a1 = l1b(2)
                c0 = l2a(2)
                c1 = l2b(2)
            end if

            if (a0 > a1) then
                tmp = a0
                a0 = a1
                a1 = tmp
            end if
            if (c0 > c1) then
                tmp = c0
                c0 = c1
                c1 = tmp
            end if

            overlap0 = max(a0, c0)
            overlap1 = min(a1, c1)

            if (overlap1 < overlap0) then
                flag = -1
            else if (overlap1 <= overlap0) then ! They touch at exactly one endpoint
                flag = 0
            else ! They overlap over a nonzero interval
                flag = 2
            end if
            return
        end if

        ! Endpoint-to-endpoint touch
        if (eq_l1al2a .or. eq_l1al2b .or. eq_l1bl2a .or. eq_l1bl2b) then
            flag = 0
            return
        end if

        ! T-junction
        if (on_segment(l1a, l1b, l2a) .or. &
            on_segment(l1a, l1b, l2b) .or. &
            on_segment(l2a, l2b, l1a) .or. &
            on_segment(l2a, l2b, l1b)) then
            flag = 4
        end if
    end function lines_intersect

end module intersections

File: src/formosa/geomorphology/geometry/intersections.py
"""
Evaluates two-dimensional geometric predicates.

This module exposes the public orientation, in-circle, and
line-segment intersection API. It dispatches calculations to the
internal Python or FORTRAN backend and selects C-interoperable
integer overloads where possible.

Last modified: 2026-08-12, En-Chi Lee (williameclee@gmail.com)
"""

import numpy as np

from formosa.utils import Backend
from formosa.geomorphology._native import intersections as intx_f
from formosa.geomorphology.geometry._backends.intersections_py import IntersectionKind
import formosa.geomorphology.geometry._backends.intersections_py as intxs_py

from numpy.typing import NDArray, ArrayLike
from formosa.utils.typing import Real, NpReal


def _point(value: ArrayLike, name: str) -> NDArray[NpReal]:
    """
    Returns a validated, real-valued 2D point.
    """
    point = np.asarray(value)
    if point.shape != (2,):
        raise ValueError(f"{name} must have shape (2,), " + f"but got {point.shape}.")
    if not np.issubdtype(point.dtype, np.number):
        raise TypeError(
            f"{name} must contain numeric coordinates, " + f"but got {point.dtype}."
        )
    if np.issubdtype(point.dtype, np.complexfloating):
        raise TypeError(f"{name} must contain real-valued coordinates.")
    return point


def _fortran_point(value: NDArray[NpReal]) -> NDArray[np.float32]:
    """
    Converts a point to the representation consumed by the FORTRAN
    backend.
    """
    return np.asarray(value, dtype=np.float32, order="F")


def _dint_dtype(points: tuple[NDArray[NpReal], ...]) -> np.dtype | None:
    """
    Selects a lossless C-interoperable integer input kind, if
    possible.
    """
    if not all(np.issubdtype(point.dtype, np.integer) for point in points):
        return None

    minimum = min(int(np.min(point)) for point in points)
    maximum = max(int(np.max(point)) for point in points)
    if np.iinfo(np.int32).min <= minimum and maximum <= np.iinfo(np.int32).max:
        return np.dtype(np.int32)
    if np.iinfo(np.int64).min <= minimum and maximum <= np.iinfo(np.int64).max:
        return np.dtype(np.int64)
    return None


def _fint_points(
    points: tuple[NDArray[NpReal], ...], dtype: np.dtype
) -> tuple[NDArray[np.integer], ...]:
    """
    Converts integer points to a C-interoperable FORTRAN
    representation.
    """
    return tuple(np.asarray(point, dtype=dtype, order="F") for point in points)


def orient_v2(
    p1: ArrayLike, p2: ArrayLike, p3: ArrayLike, backend: Backend = "fortran"
) -> Real:
    """
    Calculates the signed orientation determinant of 3 2D points.

    A positive result indicates a counterclockwise turn, a negative
    result indicates a clockwise turn, and zero indicates
    collinearity.

    Parameters
    ----------
    p1 : ArrayLike
        First real-valued coordinate with shape `(2,)`.
    p2 : ArrayLike
        Second real-valued coordinate with shape `(2,)`.
    p3 : ArrayLike
        Third real-valued coordinate with shape `(2,)`.
    backend : {"fortran", "python"}, optional
        Computational backend.
        Default backend is `"fortran"`.

    Returns
    -------
    det : int | float
        Signed orientation determinant.

        All-integer inputs return an integer when a C-interoperable
        integer kind can represent the coordinates. Other inputs
        return a float.

    Raises
    ------
    ValueError
        If a point does not have shape `(2,)` or the backend is
        unsupported.
    TypeError
        If a point contains non-numeric or complex-valued
        coordinates.

    Notes
    -----
    The Python backend uses exact integer arithmetic. The FORTRAN
    backend uses double-precision intermediates for integer inputs
    and same-kind saturating results. If a 32-bit result saturates,
    this wrapper retries the calculation using the 64-bit overload.
    The anticipated coordinate magnitudes of up to approximately
    100000 are represented exactly in double precision; integer
    coordinates above `2**53` may not be.
    """
    points = (_point(p1, "p1"), _point(p2, "p2"), _point(p3, "p3"))
    match backend:
        case "python":
            return intxs_py.orient_v2(*points)
        case "fortran":
            match _dint_dtype(points):
                case dtype if dtype == np.dtype(np.int32):
                    det = intx_f.orient_v2_int32(
                        *_fint_points(points, dtype)  # type: ignore
                    )
                    if det in (np.iinfo(np.int32).min, np.iinfo(np.int32).max):
                        det = intx_f.orient_v2_int64(
                            *_fint_points(points, np.dtype(np.int64))
                        )
                    return det
                case dtype if dtype == np.dtype(np.int64):
                    return intx_f.orient_v2_int64(*_fint_points(points, dtype))  # type: ignore
                case _:
                    return float(intx_f.orient_v2_real(*map(_fortran_point, points)))
        case _:
            raise ValueError(f"Unsupported backend {backend!r}.")


def incircle(
    a: ArrayLike,
    b: ArrayLike,
    c: ArrayLike,
    p: ArrayLike,
    oriented: bool = False,
    backend: Backend = "fortran",
) -> Real:
    """
    Calculates the signed in-circle determinant for 4 2D points.

    For counterclockwise triangle vertices, a positive result places
    `p` inside their circumcircle, zero places it on the circle, and
    a negative result places it outside. Reversing the triangle
    orientation reverses the determinant sign.

    When `oriented=True`, the result is normalised so that the
    classification is independent of the order of the triangle
    vertices.

    Parameters
    ----------
    a : ArrayLike
        First real-valued triangle coordinate with shape `(2,)`.
    b : ArrayLike
        Second real-valued triangle coordinate with shape `(2,)`.
    c : ArrayLike
        Third real-valued triangle coordinate with shape `(2,)`.
    p : ArrayLike
        Real-valued test coordinate with shape `(2,)`.
    oriented : bool, optional
        Whether to normalise the determinant sign for the
        orientation of `a`, `b`, and `c`.
        Default option is `False`.
    backend : {"fortran", "python"}, optional
        Computational backend.
        Default backend is `"fortran"`.

    Returns
    -------
    det : int | float
        Signed in-circle determinant.

        All-integer inputs return an integer when a C-interoperable
        integer kind can represent the coordinates. Other inputs
        return a float.

    Raises
    ------
    ValueError
        If a point does not have shape `(2,)`, the backend is
        unsupported, or sign normalisation is requested for a
        collinear triangle.
    TypeError
        If a point contains non-numeric or complex-valued
        coordinates.

    Notes
    -----
    The Python backend uses exact integer arithmetic. The FORTRAN
    backend uses double-precision intermediates for integer inputs
    and same-kind saturating results. If a 32-bit result saturates,
    this wrapper retries the calculation using the 64-bit overload.
    The anticipated coordinate magnitudes of up to approximately
    100000 are represented exactly in double precision; integer
    coordinates above `2**53` may not be.
    """
    points = (
        _point(a, "a"),
        _point(b, "b"),
        _point(c, "c"),
        _point(p, "p"),
    )
    match backend:
        case "python":
            det = intxs_py.incircle(*points)
        case "fortran":
            match _dint_dtype(points):
                case dtype if dtype == np.dtype(np.int32):
                    det = intx_f.incircle_int32(*_fint_points(points, dtype))  # type: ignore
                    if det in (np.iinfo(np.int32).min, np.iinfo(np.int32).max):
                        det = intx_f.incircle_int64(
                            *_fint_points(points, np.dtype(np.int64))
                        )
                    det = det
                case dtype if dtype == np.dtype(np.int64):
                    det = intx_f.incircle_int64(*_fint_points(points, dtype))  # type: ignore
                case _:
                    det = float(intx_f.incircle_real(*map(_fortran_point, points)))
        case _:
            raise ValueError(f"Unsupported backend {backend!r}.")

    if not oriented:
        return det

    orient = orient_v2(a, b, c, backend=backend)
    if orient == 0:
        raise ValueError(
            "Cannot normalise an in-circle determinant for a collinear triangle."
        )
    return det if orient > 0 else -det


def on_segment(
    a: ArrayLike, b: ArrayLike, p: ArrayLike, backend: Backend = "fortran"
) -> bool:
    """
    Determines whether a 2D point lies on a closed line segment.

    Parameters
    ----------
    a : NDArray[number]
        First point of the line segment.
        Must be a 2-by-0 real-valued coordinate.
    b : NDArray[number]
        Second point of the line segment.
        Must be a 2-by-0 real-valued coordinate.
    p : NDArray[number]
        Point to check.
    backend : {"fortran", "python"}, optional
            Computational backend.
            Default backend is `"fortran"`.

    Returns
    -------
    flag : bool
        Whether the point lies on the line segment.
    """
    points = (_point(a, "a"), _point(b, "b"), _point(p, "p"))
    match backend:
        case "python":
            return bool(intxs_py.on_segment(*points))
        case "fortran":
            return bool(intx_f.on_segment(*map(_fortran_point, points)))
        case _:
            raise ValueError(f"Unsupported backend {backend!r}.")


def bboxes_overlap(
    l1a: ArrayLike,
    l1b: ArrayLike,
    l2a: ArrayLike,
    l2b: ArrayLike,
    backend: Backend = "fortran",
) -> bool:
    """
    Determines whether 2 closed 2D segment bounding boxes overlap.

    Parameters
    ----------
    l1a : NDArray[number]
        First point of the first line segment.
        Must be a 2-by-0 real-valued coordinate.
    l1b : NDArray[number]
        Second point of the first line segment.
        Must be a 2-by-0 real-valued coordinate.
    l2a : NDArray[number]
        First point of the second line segment.
        Must be a 2-by-0 real-valued coordinate.
    l2b : NDArray[number]
        Second point of the second line segment.
        Must be a 2-by-0 real-valued coordinate.

    Returns
    -------
    flag : bool
        Whether the bounding boxes overlap.
    """
    points = (
        _point(l1a, "l1a"),
        _point(l1b, "l1b"),
        _point(l2a, "l2a"),
        _point(l2b, "l2b"),
    )
    match backend:
        case "python":
            return bool(intxs_py.bboxes_overlap(*points))
        case "fortran":
            return bool(intx_f.bboxes_overlap(*map(_fortran_point, points)))
        case _:
            raise ValueError(f"Unsupported backend {backend!r}.")


def lines_intersect_v2(
    l1a: ArrayLike,
    l1b: ArrayLike,
    l2a: ArrayLike,
    l2b: ArrayLike,
    backend: Backend = "fortran",
) -> int:
    """
    Classifies the intersection of 2 closed 2D line segments.

    The return flag has the following interpretation (as in
    :class:`IntersectionKind`):
      - `-1`: Disjoint segments
      - `0`: Endpoint contact
      - `1`: Interior 'X' crossing
      - `2`: Non-identical collinear overlap
      - `3`: Identical segments (orientation-agnostic)
      - `4`: Endpoint-on-interior 'T' junction
      - `5`: Degenerate segment (some line is actually a point)

    Parameters
    ----------
    l1a : NDArray[number]
        First point of the first line segment.
        Must be a 2-by-0 real-valued coordinate.
    l1b : NDArray[number]
        Second point of the first line segment.
        Must be a 2-by-0 real-valued coordinate.
    l2a : NDArray[number]
        First point of the second line segment.
        Must be a 2-by-0 real-valued coordinate.
    l2b : NDArray[number]
        Second point of the second line segment.
        Must be a 2-by-0 real-valued coordinate.

    Returns
    -------
    flag : int
        Intersection type.
    """
    points = (
        _point(l1a, "l1a"),
        _point(l1b, "l1b"),
        _point(l2a, "l2a"),
        _point(l2b, "l2b"),
    )
    match backend:
        case "python":
            return int(intxs_py.lines_intersect_v2(*points))
        case "fortran":
            return int(intx_f.lines_intersect_v2(*map(_fortran_point, points)))
        case _:
            raise ValueError(f"Unsupported backend {backend!r}.")
